# Predicting the fuel efficiency of a car

In [33]:
import pandas as pd
from sklearn.model_selection import train_test_split
import torch
from torch.nn.functional import one_hot
from torch.utils.data import DataLoader, TensorDataset
import torch.nn as nn

In [7]:
url = 'http://archive.ics.uci.edu/ml/machine-learning-databases/auto-mpg/auto-mpg.data'
column_names = ['MPG', 'Cylinders', 'Displacement', 'Horsepower','Weight','Acceleration','Model Year','Origin']
df = pd.read_csv(url, names=column_names, na_values="?", comment='\t', sep=" ", skipinitialspace=True)

In [8]:
df.head()

MPG  Cylinders  Displacement  Horsepower  Weight  Acceleration  \
0  18.0          8         307.0       130.0  3504.0          12.0   
1  15.0          8         350.0       165.0  3693.0          11.5   
2  18.0          8         318.0       150.0  3436.0          11.0   
3  16.0          8         304.0       150.0  3433.0          12.0   
4  17.0          8         302.0       140.0  3449.0          10.5   

   Model Year  Origin  
0          70       1  
1          70       1  
2          70       1  
3          70       1  
4          70       1

In [9]:
df.shape

(398, 8)

In [10]:
# dropping the NA rows
df = df.dropna()
df = df.reset_index(drop=True)

In [11]:
df.shape

(392, 8)

In [12]:
# Train/test split
df_train, df_test = train_test_split(df, test_size=0.2, random_state=1)

In [13]:
train_stats = df_train.describe().transpose()

In [17]:
numeric_column_names = ['Cylinders', 'Displacement', 'Horsepower', 'Weight', 'Acceleration']

In [18]:
df_train_norm, df_test_norm = df_train.copy(), df_test.copy()

In [19]:
for col_name in numeric_column_names:
    mean = train_stats.loc[col_name, 'mean']
    std = train_stats.loc[col_name, 'std']
    df_train_norm.loc[:, col_name] = (df_train_norm.loc[:, col_name] - mean) / std
    df_test_norm.loc[:, col_name] = (df_test_norm.loc[:, col_name] - mean) / std

/var/folders/2n/_8jmmysn22sbx4zjv3c2lm980000gn/T/ipykernel_68509/2894848293.py:4: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '[-0.8243028   0.3511267  -0.8243028  -0.8243028  -0.8243028   1.52655621
  0.3511267  -0.8243028  -0.8243028  -0.8243028   0.3511267  -0.8243028
  0.3511267   1.52655621  1.52655621  1.52655621  0.3511267   1.52655621
 -0.8243028   0.3511267   1.52655621 -0.8243028  -0.8243028   0.3511267
 -0.8243028  -0.8243028  -0.8243028   0.3511267  -0.8243028   1.52655621
  0.3511267  -0.8243028   0.3511267  -0.8243028  -0.8243028   1.52655621
 -0.8243028   1.52655621  1.52655621 -0.8243028  -0.8243028  -0.8243028
 -0.8243028   0.3511267  -0.8243028   1.52655621 -0.8243028  -0.8243028
  1.52655621 -0.8243028  -0.8243028  -0.8243028   1.52655621  1.52655621
  0.3511267   0.3511267   1.52655621 -0.8243028  -0.8243028   1.52655621
  1.52655621 -0.8243028  -0.8243028   0.3511267   1.52655621 -0.8243028
  

In [20]:
df_train_norm.tail()

MPG  Cylinders  Displacement  Horsepower    Weight  Acceleration  \
203  28.0  -0.824303     -0.901020   -0.736562 -0.950031      0.255202   
255  19.4   0.351127      0.413800   -0.340982  0.293190      0.548737   
72   13.0   1.526556      1.144256    0.713897  1.339617     -0.625403   
235  30.5  -0.824303     -0.891280   -1.053025 -1.072585      0.475353   
37   14.0   1.526556      1.563051    1.636916  1.470420     -1.359240   

     Model Year  Origin  
203          76       3  
255          78       1  
72           72       1  
235          77       1  
37           71       1

In [24]:
# bucketing the ModelYear into categorical variables
'''
bucket = 0 if year < 73
bucket = 1 if year >= 73 and year <=76
bucket = 2 if year >= 76 and year <= 79
bucket = 3 if year >= 79
'''


'\nbucket = 0 if year < 73\nbucket = 1 if year >= 73 and year <=76\nbucket = 2 if year >= 76 and year <= 79\nbucket = 3 if year >= 79\n'

In [26]:
boundaries = torch.tensor([73,76,79])
v = torch.tensor(df_train_norm['Model Year'].values)
df_train_norm['Model Year Bucketed'] = torch.bucketize(v, boundaries, right=True)
v = torch.tensor(df_test_norm['Model Year'].values)
df_test_norm['Model Year Bucketed'] = torch.bucketize(v, boundaries, right=True)
numeric_column_names.append('Model Year Bucketed')

In [27]:
df_train_norm.head()

MPG  Cylinders  Displacement  Horsepower    Weight  Acceleration  \
334  27.2  -0.824303     -0.530922   -0.499214 -0.555264     -0.001641   
258  18.6   0.351127      0.345625    0.186457  0.776338      1.099115   
139  29.0  -0.824303     -0.891280   -0.525586 -0.874613      0.291894   
310  37.2  -0.824303     -1.008153   -1.000281 -1.110294      0.255202   
349  33.0  -0.824303     -0.823104   -0.762934 -0.908786     -0.552019   

     Model Year  Origin  Model Year Bucketed  
334          81       1                    3  
258          78       1                    2  
139          74       2                    1  
310          80       3                    3  
349          81       2                    3

In [29]:
# one hot encoding
total_origin = len(set(df_train_norm['Origin']))
origin_encoded = one_hot(torch.from_numpy(df_train_norm['Origin'].values) % total_origin)
x_train_numeric = torch.tensor(df_train_norm[numeric_column_names].values)
x_train = torch.cat([x_train_numeric, origin_encoded], dim=1).float()
origin_encoded = one_hot(torch.from_numpy(df_test_norm['Origin'].values) % total_origin)
x_test_numeric = torch.tensor(df_test_norm[numeric_column_names].values)
x_test = torch.cat([x_test_numeric, origin_encoded], dim=1).float()

In [30]:
# creating the label tensors from the ground truth MPG values
y_train = torch.tensor(df_train_norm['MPG'].values).float()
y_test = torch.tensor(df_test_norm['MPG'].values).float()

### Training a DNN regression model

In [32]:
# dataloader with batch size 8
train_ds = TensorDataset(x_train, y_train)
batch_size = 8
torch.manual_seed(1)
train_dl = DataLoader(train_ds, batch_size=batch_size, shuffle=True)

In [34]:
# building a model with two fully connected layers where one has 8 hidden units and another has 4 hidden units
hidden_units = [8, 4]
input_size = x_train.shape[1]
all_layers = []
for hidden_unit in hidden_units:
    layer = nn.Linear(input_size, hidden_unit)
    all_layers.append(layer)
    all_layers.append(nn.ReLU())
    input_size = hidden_unit
all_layers.append(nn.Linear(hidden_units[-1], 1))
model = nn.Sequential(*all_layers)
model

Sequential(
  (0): Linear(in_features=9, out_features=8, bias=True)
  (1): ReLU()
  (2): Linear(in_features=8, out_features=4, bias=True)
  (3): ReLU()
  (4): Linear(in_features=4, out_features=1, bias=True)
)

In [35]:
# loss function and optimizer
loss_fn = nn.MSELoss()
optimizer = torch.optim.SGD(model.parameters(), lr=0.001)

In [36]:
# training the model for 200 epochs
torch.manual_seed(1)
num_epochs = 200
log_epochs = 20
for epoch in range(num_epochs):
    loss_hist_train = 0
    for x_batch, y_batch in train_dl:
        pred = model(x_batch)[:, 0]
        loss = loss_fn(pred, y_batch)
        loss.backward()
        optimizer.step()
        optimizer.zero_grad()
        loss_hist_train += loss.item()
    if epoch % log_epochs == 0:
        print(f'Epoch {epoch} Loss '
              f'{loss_hist_train/len(train_dl):.4f}')

Epoch 0 Loss 536.1047
Epoch 20 Loss 8.4361
Epoch 40 Loss 7.8695
Epoch 60 Loss 7.1891
Epoch 80 Loss 6.7064
Epoch 100 Loss 6.7603
Epoch 120 Loss 6.3107
Epoch 140 Loss 6.6884
Epoch 160 Loss 6.7549
Epoch 180 Loss 6.2029


In [37]:
# evaluating the model
with torch.no_grad():
    pred = model(x_test.float())[:, 0]
    loss = loss_fn(pred, y_test)
    print(f'Test MSE: {loss.item():.4f}')
    print(f'Test MAE: {nn.L1Loss()(pred, y_test).item():.4f}')

Test MSE: 9.5907
Test MAE: 2.1177


The MSE on the test set is 9.6, and the mean absolute error (MAE) is 2.1